$dbg does not work everywhere

### Describe the bug

Using $dbg, I noticed that there is some places where it does lead to invalid code generation and/or compilation failure.

### Reproduction Steps

```v
module main

fn fails() ?int {
	return none
}

fn noop() {
	return
}

fn main() {
	a := fails() or {
		$dbg;
		noop()
		20
	}
	println('${a}')
}
```

### Expected Behavior

The code to drop me into the debugger, like it does if moved outside the `or {}` block

### Current Behavior

bad code generation:

```sh
> v -cg run bug.v
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:17280:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t4.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:17701:4: warning: expression result unused [-Wunused-value]
  (*(string*)_t1.data);
   ^~~~~~~~~~~~~~~~~~
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:19192:4: warning: expression result unused [-Wunused-value]
  (*(os__SignalHandler*)_t3.data);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:22709:5: warning: expression result unused [-Wunused-value]
                        (*(rune*)array_pop_noscan(&r->current));
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:22908:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:23420:58: warning: format specifies type 'int' but the argument has type 'i64' (aka 'long long') [-Wformat]
                printf("%s:%-4d | %s> %s\n", ((char*)(item.file.str)), item.line, string_repeat(_SLIT(" "), i).str, item.name);
                           ~~~~                                        ^~~~~~~~~
                           %-4lld
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:23420:103: warning: format specifies type 'char *' but the argument has type 'string' (aka 'struct string') [-Wformat]
                printf("%s:%-4d | %s> %s\n", ((char*)(item.file.str)), item.line, string_repeat(_SLIT(" "), i).str, item.name);
                                      ~~                                                                            ^~~~~~~~~
/tmp/v_501/bug.01J1MYCFGE9P3NE3YGCXC39MPJ.tmp.c:23769:121: error: use of undeclared identifier 'a'
                                _MOV((v__debug__DebugContextVar[2]){{.typ=_SLIT("IError"),.value=IError_str(err)},{.typ=_SLIT("int"),.value=int_str(a)}}));
                                                                                                                                                    ^
7 warnings and 1 error generated.
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.6 df7828d

### Environment details (OS name and version, etc.)

V full version: V 0.4.6 cc14272.df7828d
OS: macos, macOS, 14.5, 23F79
Processor: 10 cpus, 64bit, little endian, Apple M1 Max

getwd: /Users/thomas/Code/github.com/ze-community/ze/main
vexe: /Users/thomas/Unix/local/v/master/v
vexe mtime: 2024-06-30 12:18:18

vroot: OK, value: /Users/thomas/Unix/local/v/master
VMODULES: OK, value: /Users/thomas/Unix/data/v/modules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.45.2
Git vroot status: weekly.2024.26-18-gdf7828d1-dirty
.git/config present: true

CC version: Apple clang version 15.0.0 (clang-1500.3.9.4)
thirdparty/tcc status: thirdparty-macos-arm64 5c1d002f
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/debug/tests/or_expr_scope.expect
+++ b/vlib/v/debug/tests/or_expr_scope.expect
@@ -1,0 +1,12 @@
+#!/usr/bin/env expect
+source "common.tcl"
+
+expect "Break on *main* in ${test_file}:7\r\n"
+
+expect "${test_file}:7 vdbg> "
+send "scope\n"
+expect "err = fail (IError)"
+
+expect "${test_file}:7 vdbg> "
+send "q\n"
+expect eof

--- a/vlib/v/debug/tests/or_expr_scope.vv
+++ b/vlib/v/debug/tests/or_expr_scope.vv
@@ -1,0 +1,10 @@
+fn r() !bool {
+	return error('fail')
+}
+
+fn main() {
+	a := r() or {
+		$dbg;
+		false
+	}
+}

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -209,6 +209,11 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 	if node.left_types.len < node.left.len {
 		g.checker_bug('node.left_types.len < node.left.len', node.pos)
 	}
+	last_curr_var_name := g.curr_var_name.clone()
+	g.curr_var_name = []
+	defer {
+		g.curr_var_name = last_curr_var_name
+	}
 
 	for i, mut left in node.left {
 		mut is_auto_heap := false
@@ -224,6 +229,7 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 		left_sym := g.table.sym(g.unwrap_generic(var_type))
 		if mut left is ast.Ident {
 			ident = left
+			g.curr_var_name << ident.name
 			// id_info := ident.var_info()
 			// var_type = id_info.typ
 			blank_assign = left.kind == .blank_ident

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -218,10 +218,11 @@ mut:
 	// where an aggregate (at least two types) is generated
 	// sum type deref needs to know which index to deref because unions take care of the correct field
 	aggregate_type_idx  int
-	arg_no_auto_deref   bool   // smartcast must not be dereferenced
-	branch_parent_pos   int    // used in BranchStmt (continue/break) for autofree stop position
-	returned_var_name   string // to detect that a var doesn't need to be freed since it's being returned
-	infix_left_var_name string // a && if expr
+	arg_no_auto_deref   bool     // smartcast must not be dereferenced
+	branch_parent_pos   int      // used in BranchStmt (continue/break) for autofree stop position
+	returned_var_name   string   // to detect that a var doesn't need to be freed since it's being returned
+	infix_left_var_name string   // a && if expr
+	curr_var_name       []string // curr var name on assignment
 	called_fn_name      string
 	timers              &util.Timers = util.get_timers()
 	force_main_console  bool // true when [console] used on fn main()
@@ -4166,6 +4167,9 @@ fn (mut g Gen) debugger_stmt(node ast.DebuggerStmt) {
 	mut count := 1
 	outer: for _, obj in scope_vars {
 		if obj.name !in vars {
+			if obj.name in g.curr_var_name {
+				continue
+			}
 			if obj is ast.Var && g.check_var_scope(obj, node.pos.pos) {
 				keys.write_string('_SLIT("${obj.name}")')
 				var_typ := if obj.ct_type_var != .no_comptime {

